src/utils.py

```diff
@@ -1515,6 +1515,75 @@ def run_hill_climbing(
     return states, actions, heuristics
 
 
+def run_policy_guided_astar(
+        initial_state: _S,
+        check_goal: Callable[[_S], bool],
+        get_valid_actions: Callable[[_S], Iterator[Tuple[_A, float]]],
+        get_next_state: Callable[[_S, _A], _S],
+        heuristic: Callable[[_S], float],
+        policy: Callable[[_S], Optional[_A]],
+        num_rollout_steps: int,
+        rollout_step_cost: float = 0,  # by default, rollout steps are free
+        max_expansions: int = 10000000,
+        max_evals: int = 10000000,
+        timeout: int = 10000000,
+        lazy_expansion: bool = False) -> Tuple[List[_S], List[_A]]:
+    """Perform A* search, but at each node, roll out a given policy for a given
+    number of time steps, creating new successors at each step.
+
+    Stop the roll out prematurely if the policy returns None.
+
+    Note that unlike the other search functions, which take get_successors as
+    input, this function takes get_valid_actions and get_next_state as two
+    separate inputs. This is necessary because we need to anticipate the next
+    state conditioned on the action output by the policy.
+
+    The get_valid_actions generates (action, cost) tuples. For policy-generated
+    transitions, the costs are ignored, and rollout_step_cost is used instead.
+    """
+
+    # Create a new successor function that rolls out the policy first.
+    def get_successors(state: _S) -> Iterator[Tuple[List[_A], _S, float]]:
+        # Get policy-based successors.
+        policy_state = state
+        policy_action_seq = []
+        for _ in range(num_rollout_steps):
+            action = policy(policy_state)
+            if action is None:
+                break
+            policy_state = get_next_state(policy_state, action)
+            policy_action_seq.append(action)
+            yield (list(policy_action_seq), policy_state, rollout_step_cost)
```

Comment:
i don't really understand the meaning of a successor here, since you're using `List[_A]` instead of the usual `_A`. is it supposed to mean, "from this state, if you take this sequence of actions in order, you'll end up at this final state"?

Reply:
yeah, I'll add that as a comment